### PR TITLE
Hcal Plan1 validation plots

### PR DESCRIPTION
**Pull request by kencall**
Validation plots are added for HcalRecHits, HcalDigis, and HcalSimhits to support Hcal Plan1 configuration. Behavior is controlled through eras customization.

### CODE REVIEW

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by hatakeyamak** [comment]
We needed to turn off the simhit vs rechit comparisons for fastsim. Still looking into what's going wrong with 2023 scenarios.

**Comment by kpedro88** [comment]
@hatakeyamak rather than turning off the comparison, perhaps you should use a customization like this:
```
from Configuration.Eras.Modifier_fastSim_cff import fastSim
fastSim.toModify( hcalRecoAnalyzer, ModuleLabel = cms.untracked.string('famosSimHits') )
```
You would, of course, also need to add a `ModuleLabel` parameter in the analyzer.
(See e.g. https://github.com/cms-sw/cmssw/blob/master/Validation/HcalHits/python/HcalSimHitStudy_cfi.py#L12-L13.)

**Comment by hatakeyamak** [comment]
Right. It should be done now.

**Comment by kencall** [comment]
@kpedro88, can you restart the test?

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by hatakeyamak** [comment]
We took a look, and we see additional DQM histograms as we wanted. So, from our point of view, this should be good to go. Can this be signed off?

**Comment by kpedro88** [comment]
@dmitrijus please check/sign

**Comment by dmitrijus** [comment]
+1

cms.untracked.string("yes")

**Comment by hatakeyamak** [comment]
We looked at the DQM outputs from above, and we are happy with the new additions. Hope to see this integrated in IB soon, which will help our next PR which we are working on now.

**Comment by abdoulline** [comment]
@davidlange6
David, may we get it merged, please?
Do that HLT could then add their stuff on top of it?

**Comment by kpedro88** [comment]
@davidlange6 ping

**Comment by davidlange6** [comment]
As I commented in the other related pr, #18353  goes first

**Comment by hatakeyamak** [comment]
I personally don't know why, because this PR and #18353 don't seem interfering/overlapping, and this has been sitting for a much longer time, but I will just wait. Thank you.

**Comment by davidlange6** [comment]
Are you not aware the hcal is currently turned off/broken in all offline workflows outside the tier0 (e.g. There is no way to test hcal Dqm changes)?

Cheers
David

On Apr 20, 2017, at 10:33 PM, Kenichi Hatakeyama <notifications@github.com<mailto:notifications@github.com>> wrote:


I personally don't know why, because this PR and #18353<https://github.com/cms-sw/cmssw/pull/18353> don't seem interfering/overlapping, and this has been sitting for a much longer time, but I will just wait. Thank you.

**Comment by kpedro88** [comment]
I thought only DQM was disabled and validation would still be run.

**Comment by davidlange6** [comment]
Yes. But we need to test the dqm changes too, no?  Anyway we are waiting on the promised fix to dqm for some time. That issue needs to be resolved (and maybe it is now..)

Cheers,
David


On 20 Apr 2017, at 23:00, Kevin Pedro <notifications@github.com<mailto:notifications@github.com>> wrote:


I thought only DQM was disabled and validation would still be run.

**Comment by hatakeyamak** [comment]
I thought only DQM was disabled and validation would still be run: right, that was my understanding. But, ok, i understand better your reasoning that you prioritize putting the DQM back in. Thank you again.

**Comment by davidlange6** [comment]
Maybe there is a bunch of code in the wrong place?

Anyway, meanwhile all the if hep17 == yes need to be changed into a Boolean (or other future proof and efficient) comparison.

Cheers,
David


On 20 Apr 2017, at 23:19, Kenichi Hatakeyama <notifications@github.com<mailto:notifications@github.com>> wrote:


I thought only DQM was disabled and validation would still be run: right, that was my understanding. But, ok, i understand better your reasoning that you prioritize putting the DQM back in. Thank you again.

**Comment by hatakeyamak** [comment]
Will work on the boolean today.

What do you mean by: Maybe there is a bunch of code in the wrong place?

**Comment by kpedro88** [comment]
There are a few DQM files modified in this PR, not just Validation - I hadn't noticed it originally.

**Comment by hatakeyamak** [comment]
DQMOffline/Hcal is primarily used for validation, while DQM uses DQM/xxx. Validation/xxx package runs only on MC, so the validation team needs DQMOffline/Hcal to have a certain set of histograms for validation using data. So, the validation team's use of DQMOffline/Hcal is expected...

**Comment by kpedro88** [comment]
@hatakeyamak somehow you included an extra commit from Sunanda...

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by hatakeyamak** [comment]
This should be good to go. The last two commits addressed the comments from @davidlange6 and also implemented a few planned fixes. It was no longer necessary to skip Validation/HcalRecHits validation for phase2 after Sunada fix PR18237.

**Comment by hatakeyamak** [comment]
@kencall looked at the DQM output from jenkins test, and we are happy with the outcome. So, from our point of view, this is good to go. Thank you.

File: Validation/HcalHits/interface/HcalSimHitStudy.h [hunk 1]
```diff
@@ -48,6 +48,22 @@ class HcalSimHitStudy: public DQMEDAnalyzer {
...
+  struct HistLim {
+    HistLim(int nbin, double mini, double maxi)
+    : n(nbin), min(mini), max(maxi) {
+    }
+    int n;
+    double min;
+    double max;
+  };
+
+  std::map<std::string, MonitorElement*> *msm_;
```
Comment (davidlange6) [inline Validation/HcalHits/interface/HcalSimHitStudy.h]: Hi @kencall - no need for this to be a pointer - if its preferred for some reason, it needs to be a unique_ptr. Thanks

**Comment by hatakeyamak** [comment]
@kencall coudln't work on this today, and this was the part I worked on initially, so I tried to fix, but obviously the rebase attempt messed things up. Now, I have to fly, so I will have to come back to this later.

**Comment by slava77** [comment]
-1

needs a rebase or a cleanup of commit history by other means

**Comment by hatakeyamak** [comment]
@davidlange6 thank you for having a close look. We simplified the simhit depth plot implementation so that we can make the same plots without the pointer you pointed out. We also use this opportunity to remove the hard-coded bin definition from the Calotower DQMOffline code (used for Validation) and now it is in sync with how we do for rechit plots. Also, added a few iphi plots which are very useful for HEP17 validation.

@slava77 I think we cleaned up the commit so this should be ready to be considered for being merged.

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by dmitrijus** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**